src/Bicep.Core/TypeSystem/TypeAssignmentVisitor.cs

```diff
@@ -1,913 +1,1138 @@
+            => AssignTypeWithCaching(syntax, () => {
+                var diagnostics = new List<Diagnostic>();
+                var reference = assignFunc(diagnostics);
+
+                return new TypeAssignment(reference, diagnostics);
+            });
+
+        private TypeSymbol? CheckForCyclicError(SyntaxBase syntax)
+        {
+            if (cyclesBySyntax.TryGetValue(syntax, out var cycle))
+            {
+                // there's a cycle. stop visiting now or we never will!
+                if (cycle.Length == 1)
+                {
+                    return UnassignableTypeSymbol.CreateErrors(DiagnosticBuilder.ForPosition(syntax).CyclicSelfReference());
+                }
+
+                var syntaxBinding = bindings[syntax];
+                
+                // show the cycle as originating from the current syntax symbol
+                var cycleSuffix = cycle.TakeWhile(x => x != syntaxBinding);
+                var cyclePrefix = cycle.Skip(cycleSuffix.Count());
+                var orderedCycle = cyclePrefix.Concat(cycleSuffix);
+
+                return UnassignableTypeSymbol.CreateErrors(DiagnosticBuilder.ForPosition(syntax).CyclicExpression(orderedCycle.Select(x => x.Name)));
+            }
+
+            return null;
+        }
+
+        public override void VisitResourceDeclarationSyntax(ResourceDeclarationSyntax syntax)
+            => AssignTypeWithDiagnostics(syntax, diagnostics => {
+                var stringSyntax = syntax.TryGetType();
+
+                if (stringSyntax != null && stringSyntax.IsInterpolated())
+                {
+                    // TODO: in the future, we can relax this check to allow interpolation with compile-time constants.
+                    // right now, codegen will still generate a format string however, which will cause problems for the type.
+                    return UnassignableTypeSymbol.CreateErrors(DiagnosticBuilder.ForPosition(syntax.Type).ResourceTypeInterpolationUnsupported());
+                }
+
+                var stringContent = stringSyntax?.TryGetLiteralValue();
+                if (stringContent == null)
+                {
+                    return UnassignableTypeSymbol.CreateErrors(DiagnosticBuilder.ForPosition(syntax.Type).InvalidResourceType());
+                }
+
+                // TODO: This needs proper namespace, type, and version resolution logic in the future
+                var typeReference = ResourceTypeReference.TryParse(stringContent);
+                if (typeReference == null)
+                {
+                    return UnassignableTypeSymbol.CreateErrors(DiagnosticBuilder.ForPosition(syntax.Type).InvalidResourceType());
+                }
+
+                var declaredType = resourceTypeProvider.GetType(typeReference);
+                
+                // just established the declared type - assign it!
+                AssignDeclaredType(syntax, declaredType);

```

Comment:
Like I mentioned in the other comments and yesterday, this approach is a bit fragile because it requires the visitor to run on the whole program to work right. I'm planning a refactor to make this more robust and consistent. It'll come on another PR.